### PR TITLE
Upgrade VMs

### PR DESCRIPTION
**Pull request by GaborApatiNagy**
Implementing Upgrading VMs to Windows Update

New screen features:
-Create skeleton of the Upgrade VM dialog
-Implement a Go To Console link (just a link that opens a vm console)
-Implement logic to show the right status for a VM (includes rules to
get the states). Includes updating the dialog whenever the status changes.
-Add status texts to Messages, proper text on the UI
-Add stub methods to VM class in XenAPI-Extensions to r/w states from
Guest Agent
-UI: Implement rules what vms to display, checkbox rules, read only vms,
and when to enable the Upgrade button
-Implement Upgrade button (writes state for GA, mounts ISO) + maximum
parallel executions/actions

Signed-off-by: Gabor Apati-Nagy <gabor.apati-nagy@citrix.com>

### CODE REVIEW

File: XenAdmin/Dialogs/UpgradeVm/UpgradeVmDialog.cs [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+                //    return VmUpgradeState.UPDATEABLE;
+
+                return VmUpgradeState.UNKNOWN;
+            }
+        }
+
+        private void ShowVmConsole(VM vm)
+        {
+            mainWindow.Invoke(delegate
+            {
```
Comment (GaborApatiNagy) [inline XenAdmin/Dialogs/UpgradeVm/UpgradeVmDialog.cs]: Code review comment: we don't think Invoke is needed here, Gabor needs to check this

File: XenAdmin/Dialogs/UpgradeVm/UpgradeVmDialog.cs [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+        private void RefreshWarning()
+        {
+
+        }
+
+        private void UpgradeVmDialog_FormClosed(object sender, FormClosedEventArgs e)
+        {
+            foreach (IXenConnection xenConnection in ConnectionsManager.XenConnectionsCopy)
+            {
+                xenConnection.Cache.DeregisterBatchCollectionChanged<VM>(VM_BatchCollectionChanged);
```
Comment (GaborApatiNagy) [inline XenAdmin/Dialogs/UpgradeVm/UpgradeVmDialog.cs]: This should be CollectionChanged

File: XenAdmin/Dialogs/UpgradeVm/UpgradeVmDialog.cs [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+            {
+                var status = VMStateHelper.GetUpgradeStateForVm(VM);
+                
+                return CAN_UPGRADE_STATES.Contains(status);
+            }
+        }
+
+        private enum VmUpgradeState 
+        {
+            UNKNOWN = 0,
```
Comment (GaborApatiNagy) [inline XenAdmin/Dialogs/UpgradeVm/UpgradeVmDialog.cs]: code review comment: numbers are not needed here

**Comment by MihaelaStoica** [comment]
Looks fine. Merging.

**Pull request merged**